### PR TITLE
Merge Flambda 2 .cmxa/.a files into new ocamloptcomp.cmxa/.a

### PR DESCRIPTION
**Pull request by mshinwell**
This will avoid clients of compilerlibs having to explicitly link against the Flambda 2 archives.  There is currently no tool for merging `.cmxa` files as far as I know, so I wrote one.  This may need to be extended in the future to handle `.cma` files too.

### CODE REVIEW

File: tools/dune [hunk 1]
```diff
@@ -48,3 +48,10 @@
   )
   (section bin)
   (package ocaml))
+
+(executable
+  (name merge_archives)
+  (modes native)
+  (flags -nostdlib)
```
Comment (xclerc) [inline tools/dune:55]: (Why is this needed? the code clearly shows that the stdlib is actually used,
presumably transitively from the other dependencies.)

File: tools/merge_archives.ml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+let merge_cma ~target:_ ~archives:_ =
+  failwith "not yet implemented"
+
+let read_cmxa filename =
+  let chan = open_in_bin filename in
+  let magic =
+    really_input_string chan (String.length Config.cmxa_magic_number)
+  in
+  let (cmxa : Cmx_format.library_infos) = input_value chan in
```
Comment (xclerc) [inline tools/merge_archives.ml:25]: It looks like we should check the magic number here;
there is a consistency check in the function below, but
my fear is the unmarshaling of something with the wrong
shape (and the segfault that could follow).

Or is it written this way because you would like to be
compatible across versions?

File: tools/merge_archives.ml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  close_in chan;
+  magic, cmxa
+
+let merge_cmxa0 ~archives =
+  let magic_and_cmxa_list = List.map read_cmxa archives in
+  let magics = List.map fst magic_and_cmxa_list in
+  let cmxa_list = List.map snd magic_and_cmxa_list in
+  let magic =
+    match String_set.elements (String_set.of_list magics) with
```
Comment (xclerc) [inline tools/merge_archives.ml:34]: (If the check is added in the function above, this one can
go, and indeed `read_cmxa` does not need to return the
couple.)

File: tools/merge_archives.ml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  magic, cmxa
+
+let merge_cmxa ~target ~archives =
+  let magic, cmxa = merge_cmxa0 ~archives in
+  let chan = open_out_bin target in
+  output_string chan magic;
+  output_value chan cmxa;
+  close_out chan
+
+let has_extension archive ~ext =
```
Comment (xclerc) [inline tools/merge_archives.ml:77]: (This looks similar to `Filename.check_suffix`.)

File: tools/merge_dot_a_files.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+for archive in $archives; do
+  archive=$(absolute_path $archive)
+  ar t $archive | grep '\.o$' >> $files
+  (cd $tempdir && ar x $archive)
+done
+
+cd $tempdir
+rm -f $target
+
+ar r $target $(cat $files)
```
Comment (xclerc) [inline tools/merge_dot_a_files.sh:41]: With my other hat, I would add a `D` (*i.e.* `ar rD` ...) to make it deterministic.

File: tools/merge_dot_a_files.sh [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+rm -rf $tempdir
```
Comment (xclerc) [inline tools/merge_dot_a_files.sh:43]: (I would possibly use a trap, but I doubt we really care.)

**Comment by xclerc** [comment]
I understand this is an internal tool for power users, so I did not
pay attention to *e.g.* the order of `ccobjs`/`ccopts`, packs, and
other complications.

File: tools/dune (hunk 1 above)
Comment (mshinwell) [reply tools/dune]: I was following the stanzas for the other executables, but in fact it seems not to be needed.  Removed.

File: tools/merge_archives.ml (hunk 2 above)
Comment (mshinwell) [reply tools/merge_archives.ml:25]: It deliberately doesn't check the magic number so it can be used across versions (it used to, and was problematic).

**Comment by mshinwell** [comment]
Regarding ordering: the `.cmxa` tool aims to preserve the ordering provided on the command line, in particular in case C compiler options are present where there is an ordering constraint.  The script for merging `.a` files does preserve the order, which is important there.

File: tools/merge_dot_a_files.sh (hunk 5 above)
Comment (mshinwell) [reply tools/merge_dot_a_files.sh:41]: Fixed

File: tools/merge_archives.ml (hunk 3 above)
Comment (mshinwell) [reply tools/merge_archives.ml:34]: Now not applicable I think

File: tools/merge_archives.ml (hunk 4 above)
Comment (mshinwell) [reply tools/merge_archives.ml:77]: Fixed to use that function

**Comment by xclerc** [comment]
I was a tiny bit worried about `lib_ccobjs` because it goes through a set,
but it also looks fine for the current use case.

**Review by xclerc — APPROVED** [review]

File: tools/merge_dot_a_files.sh (hunk 6 above)
Comment (mshinwell) [reply tools/merge_dot_a_files.sh]: I think I've fixed this, please check it looks ok.

**Comment by mshinwell** [comment]
`lib_ccobjs` deliberately uses a set, because it seems fine to dedup required object files, but it might not be fine to provide them duplicated to a linker (iirc).  Although maybe if the user had provided all the `.a` files anyway (without using this tool) they would have ended up with the same error...

File: tools/merge_dot_a_files.sh (hunk 6 above)
Comment (xclerc) [reply tools/merge_dot_a_files.sh]: Looks fine.

File: tools/merge_archives.ml (hunk 2 above)
Comment (xclerc) [reply tools/merge_archives.ml:25]: I think it has the potential of failing badly, but again this is for
power users so we can let them shoot themselves in the foot
since they are powerful enough to recover.

File: tools/merge_archives.ml (hunk 3 above)
Comment (xclerc) [reply tools/merge_archives.ml:34]: I concur.

File: tools/merge_archives.ml (hunk 2 above)
Comment (mshinwell) [reply tools/merge_archives.ml:25]: Ah, I see, because you mean that "library_infos" might have changed.  I think this is probably ok for an internal build system tool.

Comment (xclerc) [reply tools/merge_archives.ml:25]: Yes, that type or any one transitively used from it.
In the end I agree it is fine; the tool is simple enough so that
if we see an unexplained failure, it will come from the unmarshaling.
